### PR TITLE
EDM-632: device/applications: add embedded compose support

### PR DESCRIPTION
**Pull request by hexfusion**
This PR adds embedded compose application support.

Embeded applications are expected to be written to a sub directory of `/usr/local/etc/compose/manifests` where the sub directory is the application's name.

```
/usr/local/etc/compose/manifests/embedded-demo-app/podman-compose.yaml
```

The agent will start the application in the same way it handles image based applications and will also report its status. Below is an embedded application `embedded-demo-app` along side a declared OCI image-based application `compose-oci-example` in status.

```
bin/flightctl get devices/aupiu13663re92hg9q61cajan83gjp8e6tubpa01ncq9a0kgu6eg -o json | jq .status.applications
[
  {
    "name": "embedded-demo-app",
    "ready": "3/3",
    "restarts": 0,
    "status": "Running"
  },
  {
    "name": "compose-oci-example",
    "ready": "3/3",
    "restarts": 0,
    "status": "Running"
  },
]
```

This PR also adds a shutdown manager which greatly improves graceful termination and ensures draining applications before shutdown. This is very important for embedded apps because if we don't drain them they will potentially start on reboot even if removed from the image.

But also in general we want to ensure we gracefully terminate all applications. A similar approach is used with the kubelet.

```
time="2024-11-07T16:57:55.501666Z" level=info msg="Agent received shutdown signal: interrupt"
time="2024-11-07T16:57:55.501711Z" level=info msg="Shutting down: applications"
time="2024-11-07T16:57:55.501725Z" level=info msg="Stopping podman monitor"
time="2024-11-07T16:57:55.502417Z" level=info msg="Draining 1 applications"
time="2024-11-07T16:58:08.682052Z" level=info msg="Removed network a3afb4b240d1d1bcec0e8efe77ecb29d77d6b22ffe3b78cd76fac5c7defd53ce"
time="2024-11-07T16:58:08.686757Z" level=info msg="Podman drain complete"
time="2024-11-07T16:58:08.687210Z" level=info msg="Podman monitor stopped!!"
time="2024-11-07T16:58:08.687343Z" level=info msg="Shutting down: agent"
time="2024-11-07T16:58:08.687413Z" level=info msg="Shutdown complete in 13.185708161s"
time="2024-11-07T16:58:08.687467Z" level=info msg="Hooks manager: context closed"
time="2024-11-07T16:58:08.687551Z" level=info msg="Disk monitor stopped"
time="2024-11-07T16:58:08.688072Z" level=info msg="CPU monitor stopped"
time="2024-11-07T16:58:08.688149Z" level=info msg="Agent shutdown complete"
```

### CODE REVIEW

**Review by celdrake — COMMENTED** [review]
LGTM, The agent reports the embedded application and the API returns it as other apps. The UI shows the status correctly.

File: internal/agent/device/applications/controller.go [hunk 1]
```diff
@@ -50,9 +51,8 @@ func (c *Controller) Sync(ctx context.Context, current, desired *v1alpha1.Render
 		return err
 	}
 
-	// if this is the steady state, only ensure apps
-	if !spec.IsUpgrading(current, desired) {
```
Comment (hexfusion) [inline internal/agent/device/applications/controller.go:54]: note: updated to ensure that we only have one idempotent flow

File: internal/agent/device/applications/controller.go [hunk 2]
```diff
@@ -64,32 +64,30 @@ func (c *Controller) Sync(ctx context.Context, current, desired *v1alpha1.Render
...
 			return err
 		}
 		if err := c.manager.Remove(app); err != nil {
 			return err
 		}
-		c.log.Infof("Removed application %s", app.Name())
 	}
 
-	for _, app := range added {
+	for _, app := range diff.Ensure {
```
Comment (hexfusion) [inline internal/agent/device/applications/controller.go:81]: note: ensure now covers added and unchanged apps.

File: internal/agent/device/applications/lifecycle/compose.go [hunk 3]
```diff
@@ -28,81 +29,106 @@ func NewCompose(log *log.PrefixLogger, podman *client.Podman) *Compose {
...
 
 func (c *Compose) remove(ctx context.Context, action *Action) error {
+	c.log.Debugf("Removing application %s", action.Name)
+
 	// by using podman directly we can avoid the need to parse the compose file.
 	// this makes the reconciliation process faster and more reliable as the
 	// compose file is not required.
 	labels := []string{fmt.Sprintf("com.docker.compose.project=%s", action.Name)}
 
+	var errs []error
```
Comment (hexfusion) [inline internal/agent/device/applications/lifecycle/compose.go:55]: note: these actions are important so lets not bail because one fails.

**Review by DakCrowder — APPROVED** [review]

**Pull request merged**